pkg/ebpf/telemetry/errors_telemetry.go

```diff
@@ -43,213 +40,346 @@ var helperNames = map[int]string{
-	b.mtx.Lock()
-	defer b.mtx.Unlock()
+// initializeProbeKeys assignes an integer key to each probe in the manager.
+// This key is the index in the array at which telemetry for this probe is kept
+// See pkg/ebpf/c/telemetry_types.h for the definition of the struct holding telemetry
+func initializeProbeKeys(programs map[string]*ebpf.ProgramSpec, bpfTelemetry *EBPFTelemetry) {
+	bpfTelemetry.mtx.Lock()
+	defer bpfTelemetry.mtx.Unlock()
 
-	// first manager to call will populate the maps
-	if b.mapErrMap == nil {
-		b.mapErrMap, _ = maps.GetMap[uint64, MapErrTelemetry](m, probes.MapErrTelemetryMap)
-	}
-	if b.helperErrMap == nil {
-		b.helperErrMap, _ = maps.GetMap[uint64, HelperErrTelemetry](m, probes.HelperErrTelemetryMap)
+	for fn := range programs {
+		bpfTelemetry.probeKeys[fn] = bpfTelemetry.probeIndex
+		bpfTelemetry.probeIndex++
 	}
+}
 
-	if err := b.initializeMapErrTelemetryMap(m.Maps); err != nil {
-		return err
-	}
-	if err := b.initializeHelperErrTelemetryMap(); err != nil {
-		return err
+func patchConstant(ins asm.Instructions, symbol string, constant int64) error {
+	indices := ins.ReferenceOffsets()[symbol]
+
+	ldDWImm := asm.LoadImmOp(asm.DWord)
+	for _, index := range indices {
+		load := &ins[index]
+		if load.OpCode != ldDWImm {
+			return fmt.Errorf("symbol %v: load: found %v instead of %v", symbol, load.OpCode, ldDWImm)
+		}
+		load.Constant = constant
 	}
+
 	return nil
 }
 
-func (b *EBPFTelemetry) initializeMapErrTelemetryMap(maps []*manager.Map) error {
-	if b.mapErrMap == nil {
-		return nil
+func patchEBPFInstrumentation(m *manager.Manager, bpfTelemetry *EBPFTelemetry, bytecode io.ReaderAt, shouldSkip func(string) bool) error {
+	programs, err := m.GetProgramSpecs()
+	if err != nil {
+		return fmt.Errorf("unable to get program specs: %w", err)
 	}
 
-	z := new(MapErrTelemetry)
-	h := keyHash()
-	for _, m := range maps {
-		// Some maps, such as the telemetry maps, are
-		// redefined in multiple programs.
-		if _, ok := b.mapKeys[m.Name]; ok {
+	return PatchEBPFInstrumentation(programs, bpfTelemetry, bytecode, shouldSkip)
+}
+
+// PatchEBPFInstrumentation accepts eBPF bytecode and patches in the eBPF instrumentation
+func PatchEBPFInstrumentation(programs map[string]*ebpf.ProgramSpec, bpfTelemetry *EBPFTelemetry, bytecode io.ReaderAt, shouldSkip func(string) bool) error {
+	initializeProbeKeys(programs, bpfTelemetry)
+
+	sizes, err := parseStackSizesSections(bytecode, programs)
+	if err != nil {
+		return fmt.Errorf("failed to parse '.stack_sizes' section in file: %w", err)
+	}
+
+	for fn, p := range programs {
+		space := sizes.stackHas8BytesFree(fn)
+		// return error if there is not enough stack space to cache the pointer to the telemetry array
+		// if the program is intended to be skipped then we continue since we still need to patch a nop
+		if !space && !shouldSkip(p.Name) {
+			return fmt.Errorf("Function %s does not have enough free stack space for instrumentation", fn)
```

Comment:
Is there any way we could fail at build time if there isn't enough stack space? That would simplify things and I think prevent the issue with passing the `io.ReaderAt` around. We would need to ensure we check on runtime compilation builds as well.

Reply:
We can. But then we would fail for functions that will not be instrumented as well.

Reply:
Couldn't we ignore those functions during the build-time check too?